src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Instance/Authenticator.cs

```diff
@@ -109,18 +109,24 @@ public async Task UpdateFromTemplateAsync(RequestContext requestContext)
                 var authorityUri = new Uri(Authority);
                 var host = authorityUri.Host;
 
-                // The authority could be https://{AzureAD host name}/{tenantid} OR https://{Dsts host name}/dstsv2/{tenantid}
+                // The authority could be https://{AzureAD host name}/{tenantid} OR https://{Dsts host name}/dstsv2/{tenantid} OR
+                // https://{adfs host name}/adfs/{tenantid}
```

Comment:
>https://{adfs host name}/adfs/{tenantid} [](start = 19, length = 40)

I thought ADFS wouldn't necessarily have a tenantid. we probably have to support both?

Reply:
ADFS will never have a tenantId. ARM etc. will advertise and use /adfs for the OIDC endpoint which is correct and works. KeyVault has code patterns from AAD where they take a "base" authority URI in their code and add a tenantId as a suffix to that, so they end-up using /adfs/tenantId in the endpoints. They can't fix the code to directly support the ADFS authority URI so the ask is for ADAL to support passing-in the invalid authority URI and correcting it under-the-hood. Note that this same ask will be levied against MSAL as well. It seems there is a lot of this code from Azure Services that isn't going to be precise about the OIDC protocol and how authority URIs work, so it may be reasonable to have ADAL and MSAL "tolerate" such code and internally look for "/adfs" on the first segment and make the necessary transformations or w/e for ADFS. In #1653 we discussed this a bit more and how the MSAL design for .WithAdfsAuthority is also problematic for these same kinds of code patterns. Web applications that integrate with identity need to be built supporting both identity providers, but are we going to make end-users parse the authority URI again themselves in their code and branch to call .WithAuthority or .WithAdfsAuthority? This means ADAL / MSAL internally understanding authority URIs that begin with /adfs and "knowing" that it is an ADFS-backed IDP and whatever that needs to mean to make things work.

Reply:
sounds good. thanks for clarifying. I will remove the {tenantid} here, but it doesn't impact any of the changes.

---
In reply to: [378542753](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1694#discussion_r378542753) [](ancestors = 378542753)

Reply:
nice, as discussed. we should have tests for both scenarios.

Reply:
yes...will add.